### DISCUSSION TITLE
Question:  Changing hardcoded setup views/viewmodels

### DISCUSSION THREAD

**Discussion by Allann** [OrchardCMS/OrchardCore#15851] [Q&A]
I'm very new to Orchard eco system.  Is there a simple way to change the wording and add additional fields to setup module without completely replacing OrchardCore.Setup?  I notice there are many hardcoded values in the views and was wondering how these would be replaced in this modular system?

I created a new Setup but now there are 2 when creating tenants.  Not sure if I have approached this correctly.  I couldn't find any documentation on this either, but I may not have been searching for the correct things.

**Comment by hishamco** [comment]
AFAIK you should modify the Setup module, but there's notion of customizing fields. BTW this question is asked for the second time. Hopefully I will find a time to experiment such field customization

**Comment by Allann** [comment]
I should modify the existing OrchardCore.Setup, or create my own?  If existing that seems a little backward because i thought we should be importing NuGet packages, not building everything from scratch.  If create my own, how do I make the Shell call mine?
Customised fields sounds interesting, is there any doco on that?  I found about customising field templates, not adding new fields.

**Comment by hishamco** [comment]
> how do I make the Shell call mine?

/cc @jtkech

**Comment by Allann** [comment]
just as an aside, I added a new Admin too, and make no reference in the recipes to OrchardCore.Admin, yet I end up with both.  This ends up creating 2 admin menus.

I notice that the AdminFilter has the following method:
```
        private bool IsNameAdmin(ActionExecutingContext context)
        {
            return string.Equals(context.Controller.GetType().Name, "Admin", StringComparison.OrdinalIgnoreCase);
        }
```
This uses only the name and therefore will respond to all modules called Admin regardless of which namespace the controller is in.  Shouldn't it use `FullName` instead?

**Comment by jtkech** [comment]
Regarding the ability to override the setup views.

we can override any view (shapes and pure mvc views) through themes but during a setup there is no active themes. We now share the views between modules that are under a `Views/Shared` folder. So, because the app is also a module (and a super theme), i tried to define these views at the app level.

- YourApp
    - Views
        - Shared
            - Index.cshtml

        - _Layout.cshtml
        - _ViewImports.cshtml

It works but then any index action of any controller targets this view, so this is not a good solution.

Would be better to define the view under

- YourApp
    - Views
        - OrchardCore.Setup
            - Setup
                - Index.cshtml

We have another views location expander that searches for this, but again it is only used when the theming engine is there, which is no the case before a setup.

Maybe, the one that is always there but that only searches `Shared/{0}.cshtml` ({0} = action), would have to also search for `Shared/{2}/{1}/{0}.cshtml` ({2} = area = module, {1} = controller).

I will think about it

**Comment by Allann** [comment]
thanks.  I'll give those options a go and see how I go.  I would much prefer to use features that come out-the-box since customisations are harder to upgrade going forward.

Actually, due to being new I mistook the "Theme" as being a web theme, not an initial site builder.  The documentation does not make that clear.  I have started creating a theme to help build the initial site for tenants and this is working so much easier, and no longer need the customisations as I can build these into the theme.

Your mention that views can be overwritten in themes gave me the push in the right direction.  I think this needs to be a lot more prevalent in the How to guides, and/or at least highlighted in the "Creating a website" sections.

**Comment by sebastienros** [comment]
> can be overwritten in themes

And also from the main web project too
